flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergSink.java

```diff
@@ -0,0 +1,696 @@
+      @Nullable String uidPrefix,
+      Map<String, String> snapshotProperties,
+      boolean upsertMode,
+      boolean overwrite,
+      TableSchema tableSchema,
+      DistributionMode distributionMode,
+      int workerPoolSize,
+      long targetDataFileSize,
+      FileFormat dataFileFormat) {
+    Preconditions.checkNotNull(tableLoader, "Table loader shouldn't be null");
+    Preconditions.checkNotNull(table, "Table shouldn't be null");
+
+    this.tableLoader = tableLoader;
+    this.table = table;
+    this.equalityFieldIds = equalityFieldIds;
+    this.equalityFieldColumns = equalityFieldColumns;
+    this.uidPrefix = uidPrefix == null ? "" : uidPrefix;
+    this.snapshotProperties = snapshotProperties;
+    this.flinkRowType = toFlinkRowType(table.schema(), tableSchema);
+    this.upsertMode = upsertMode;
+    this.overwrite = overwrite;
+    this.distributionMode = distributionMode;
+    this.workerPoolSize = workerPoolSize;
+    this.targetDataFileSize = targetDataFileSize;
+    this.dataFileFormat = dataFileFormat;
+  }
+
+  @Override
+  public DataStream<RowData> addPreWriteTopology(DataStream<RowData> inputDataStream) {
+    return distributeDataStream(
+        inputDataStream,
+        equalityFieldIds,
+        table.spec(),
+        table.schema(),
+        flinkRowType,
+        distributionMode,
+        equalityFieldColumns);
+  }
+
+  @Override
+  public StreamWriter createWriter(InitContext context) {
+    return restoreWriter(context, null);
+  }
+
+  @Override
+  public StreamWriter restoreWriter(
+      InitContext context, Collection<StreamWriterState> recoveredState) {
+    RowDataTaskWriterFactory taskWriterFactory =
+        new RowDataTaskWriterFactory(
+            SerializableTable.copyOf(table),
+            flinkRowType,
+            targetDataFileSize,
+            dataFileFormat,
+            equalityFieldIds,
+            upsertMode);
+    StreamWriter streamWriter =
+        new StreamWriter(
+            table.name(),
+            taskWriterFactory,
+            context.getSubtaskId(),
+            context.getNumberOfParallelSubtasks());
+    if (recoveredState == null) {
+      return streamWriter;
+    }
+
+    return streamWriter.restoreWriter(recoveredState);
+  }
+
+  @Override
+  public SimpleVersionedSerializer<StreamWriterState> getWriterStateSerializer() {
+    return new StreamWriterStateSerializer();
+  }
+
+  @Override
+  public DataStream<CommittableMessage<FilesCommittable>> addPreCommitTopology(
+      DataStream<CommittableMessage<FilesCommittable>> writeResults) {
+    return writeResults
+        .map(
+            new RichMapFunction<
+                CommittableMessage<FilesCommittable>, CommittableMessage<FilesCommittable>>() {
+              @Override
+              public CommittableMessage<FilesCommittable> map(
+                  CommittableMessage<FilesCommittable> message) {
+                if (message instanceof CommittableWithLineage) {
```

Comment:
this if check is a little weird. The problem seems to be that `CommittableMessage` interface doesn't expose `getCommittable`, which I think we probably should add in Flink upstream.

It is also weird that Flink interface has the `CommT` generic type, but it is not used or exposed in the interface.
```
public interface CommittableMessage<CommT> {
    /** The subtask that created this committable. */
    int getSubtaskId();

    /**
     * Returns the checkpoint id or empty if the message does not belong to a checkpoint. In that
     * case, the committable was created at the end of input (e.g., in batch mode).
     */
    OptionalLong getCheckpointId();
}
```

Reply:
In fact, if we can directly getcheckpointid\jobid\subtaskId in `prepareCommit`, we do not need to assign them here.

The following is the Committable design document for reference.
>We also introduce two common wrappers around the Sink Committables to ease the implementation of the topologies based on committables.
![image](https://user-images.githubusercontent.com/59213263/185310617-ace39931-86f5-4b01-8acc-faa12d458044.png)
>In these topologies, the committable will always be part of the CommittableWithLinage to have a notion to which checkpoint it belongs and from which subtask it was sent. In this case, it can be either the SinkWriter or Committer depending on which topology is used (pre- or post-commit). Moreover, we introduce the CommittableSummary that is sent once as the last message of this checkpoint. Down-stream operators receiving this message can be sure that the subtask the message was coming from does not send more messages in this checkpoint.